### PR TITLE
(PW-3251) Processing-of-File-format-messages

### PR DESCRIPTION
**Pull request by malai-subra**
## Summary

This PR fixes parsing of file-format MX messages (PW-3251) that have `AppHdr` as the XML root or bare `AppHdr + Document` without a single wrapper element.

### Root causes fixed

1. **`MxSwiftMessage` / `MxParseUtils`**: XML with `AppHdr` as the document root caused "Illegal to have multiple roots" errors in the SAX-based metadata parser. Fixed by wrapping in `<RequestPayload>` via `wrapIfAppHdrRoot()` before DOM-based (`MxNode`) and SAX-based parsing.

2. **`AbstractMX.parse()`** (`MxReadImpl`): XML where the `Document` element carries an undeclared namespace prefix (e.g. `<ns2:Document>` with no `xmlns:ns2` declaration) caused SAX to reject the document before JAXB could unmarshal it. Fixed by `stripUndeclaredDocumentPrefix()` which strips the prefix from element tags when no declaration exists.

3. **`MxSwiftMessage.getAppHdr()`**: `AppHdrParser` uses SAX internally and also failed on bare `AppHdr + Document` XML. Fixed by applying `wrapIfAppHdrRoot()` in `getAppHdr()` so sender/receiver metadata is correctly extracted from the header.

### All entry points covered

- `new MxSwiftMessage(String xml)` ✓
- `new MxSwiftMessage(InputStream)` ✓
- `new MxSwiftMessage(File)` / `MxSwiftMessage.parse(File)` ✓
- `AbstractMX.parse(String xml)` ✓

### CODE REVIEW

File: iso20022-core/src/main/java/com/prowidesoftware/swift/model/mx/MxParseUtils.java [hunk 1]
```diff
@@ -316,6 +322,54 @@ private static Optional<MxId> enrichBusinessService(MxId mxId, final String xml)
...
+     *
+     * <p>If the prefix is declared or no prefixed Document element is found, the XML is returned unchanged.
+     *
+     * @param xml original XML content
+     * @return XML with the undeclared document prefix stripped from element tags, or the original if not needed
+     * @since 10.3.10
+     */
+    public static String stripUndeclaredDocumentPrefix(String xml) {
+        if (xml == null) return null;
+        Matcher m = Pattern.compile("<([a-zA-Z_][\\w.-]*):Document[\\s>/]").matcher(xml);
```
Comment (zubri) [inline iso20022-core/src/main/java/com/prowidesoftware/swift/model/mx/MxParseUtils.java]: **Pattern compiled on every parse (hot path).** This `Pattern.compile("<([a-zA-Z_][\\w.-]*):Document[\\s>/]")` runs on every invocation of `stripUndeclaredDocumentPrefix()`, which is now on the main parse path — `MxReadImpl.parse()` for *every* message, plus `MxSwiftMessage.extractMetadata()`/`updateMetadata()`. So a regex is recompiled per parse even for well-formed messages that need no stripping. The sibling `APPHDR_ROOT_PATTERN` (and the pre-existing `pattern`) are already `private static final` constants — this one should be too. The inner declaration-check pattern below only compiles when a prefixed `Document` is actually found, so that one is fine to leave inline.

File: iso20022-core/src/main/java/com/prowidesoftware/swift/model/mx/MxParseUtils.java [hunk 2]
```diff
@@ -316,6 +322,54 @@ private static Optional<MxId> enrichBusinessService(MxId mxId, final String xml)
...
+        if (!m.find()) return xml;
+        String prefix = m.group(1);
+        if (Pattern.compile("xmlns:" + Pattern.quote(prefix) + "\\s*=")
```
Comment (zubri) [inline iso20022-core/src/main/java/com/prowidesoftware/swift/model/mx/MxParseUtils.java]: **Undeclared-prefix detection can misfire when the prefix is declared elsewhere.** The "is it declared?" check searches for `xmlns:<prefix>=` across the *entire* document. If the `Document` prefix (e.g. `ns2`) is undeclared on `Document` but the same prefix string happens to be declared in an unrelated subtree (supplementary/signature data, etc.), this returns the XML unchanged and the `Document` still fails SAX parsing — the targeted bug is silently not fixed. This is a missed-fix rather than a regression, but consider scoping the check to the `Document` element's own start tag, or at least documenting the limitation in the Javadoc.

File: iso20022-core/src/test/java/com/prowidesoftware/issues/IssueJira3251Test.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            MxSwiftMessage msg = new MxSwiftMessage(xml);
+            assertNotNull(msg);
+            assertEquals("semt.002.001.11", msg.getIdentifier());
+            assertEquals("TRGTXE2SXXX", msg.getSender());
+            assertEquals("EMCFNL2AXXX", msg.getReceiver());
+        } catch (Exception e) {
```
Comment (zubri) [inline iso20022-core/src/test/java/com/prowidesoftware/issues/IssueJira3251Test.java]: **Test swallows the failure cause.** Wrapping the whole body in `try { … } catch (Exception e) { fail(); }` means a parsing regression surfaces as a bare `fail()` with no message and no stack trace, defeating the diagnostic value of the test. Prefer letting the exception propagate (remove the try/catch — an uncaught exception already fails the test) or use `assertDoesNotThrow(() -> { … })`, or at minimum `fail(e.getMessage(), e)` to preserve the cause.

**Review by zubri — COMMENTED** [review]
Automated code review (medium effort). Three findings below — one efficiency issue on the parse hot path, one edge case in the undeclared-prefix detection, and a test-quality issue. The overall approach is sound and the normal single-`Document` parse path is left untouched.

**Comment by zubri** [comment]
Addressed review findings in three iterations:

- 3e7e53e2d — correctness rework: XML declaration/BOM/leading comments are kept outside the synthetic `RequestPayload` wrapper (wrapping them produced malformed XML and regressed declaration-carrying headers); prefix stripping is now structure-aware (element tags only — CDATA/comment content is never modified, and the `AppHdr` prefix is covered as well); normalization applied consistently across all entry points: generated `Mx*.parse()`, `AbstractMX.parse()`, `AppHdrParser`, and type autodetection.
- 3b630e8b4 — each entry point normalizes exactly once (internal `parseNormalized` chain removes the redundant re-scans); DOCTYPE internal-subset limitation documented in the markup scanner.
- 3eb90232f — zero-copy: the wrapper is composed virtually through `LenientPayload`/`ConcatReader` readers for the SAX, StAX and `MxNode` DOM paths, so file-format payloads are parsed without a payload-sized copy; `makeXmlLenient` returns the same instance when the declaration needs no fix; the wrap now applies only to genuinely multi-root content, so a valid standalone `AppHdr` document keeps `AppHdr` as tree root.

Full iso20022-core suite green (402 tests). Deferred, if profiling ever justifies them: SAX-filter-based prefix stripping and streaming the File input without the intermediate String.

**Review by zubri — APPROVED** [review]

**Comment by zubri** [comment]
Follow-up review pass addressed in 363bf6c69: replaced the AppHdr root-detection regex with an iterative prolog scan, fixing a potential StackOverflowError on comment-heavy prologs and a full-payload backtracking scan that regressed parsing of Document-rooted messages with an XML declaration (measured +43-67% on multi-MB payloads, now back to baseline). Also documented the identifyMessage blank-input contract change in the CHANGELOG and closed test gaps: file-format payloads via the MxSwiftMessage InputStream/File constructors, raw message preserved without the synthetic wrapper, ConcatReader unit tests, DataPDU envelope pass-through, prefix-substring collision, and prolog stress guards. Full suite green (418 tests).
